packages/@punchcard/shape/lib/function.ts

```diff
@@ -0,0 +1,30 @@
+import { Shape } from './shape';
+
+export interface FunctionArgs {
+  [argName: string]: Shape;
+}
+
+export interface FunctionShapeProps<Args extends FunctionArgs, Returns extends Shape> {
+  args: Args;
+  returns: Returns;
+}
+
+export class FunctionShape<Args extends FunctionArgs, Returns extends Shape> extends Shape {
+  public readonly Kind: 'functionShape' = 'functionShape';
+  public readonly FQN: string;
+
+  constructor(
+    public readonly args: Args,
+    public readonly returns: Returns
+  ) {
+    super();
+    // compute the FQN
+    this.FQN = `(${Object.entries(args).map(([k, v]) => `${k}: ${v.FQN}`).join(',')}) => ${returns.FQN}`;
```

Comment:
In the next version of shapes, can the FQN be a trait? Not all shapes use cases will require an FQN... and although it's nice that the user doesn't need to assign one themselves, it doesn't feel to me like it should be built in. I'm wondering what you think.

```ts
import * as s from "shapes";

class MyRecord extends s.createVRecord({}) {}

class MyFQRecord1 extends MyRecord.with({ fqn: "first" }) {}
class MyFQRecord2 extends MyFQRecord1.with({ fqn: "second" }) {} // overrides "first"
```

Semi-related: when you extend a record shape, does the result carry the same FQN? Glancing through [the source](https://github.com/punchcard/punchcard/blob/master/packages/%40punchcard/shape/lib/record.ts#L251).